[l2] Fix ensure_l2_flow: BigIPRestClient needed, not just hostname

BenjaminLudwigSAP opened the pull request:
This commit fixes two things introduced in #285:

1) Since the L2 creation flow was changed to run on both hosts simultaneously, one future now handles two devices. This means that iterating over the futures can yield two hostnames instead of one. This breaks the `octavia_l2_failed_futures` metric, which only takes one hostname as label. This commit fixes this by treating all values of the futures dictionary as lists and (in the case of a failure) iterating over them to set the `octavia_l2_failed_futures` metric for each host therein.

2) After the futures have been iterated over, the failed BigIPs are checked to see whether all BigIP hosts/guest flows failed. This check needs `BigIPRestClient` instances (to compare to `self._bigips` and `self.vcmps`, respectively), but for the BigIP guests only the hostnames are provided in the futures list.

BenjaminLudwigSAP committed:
[l2] Fix ensure_l2_flow: BigIPRestClient needed, not just hostname

This commit fixes two things introduced in #285:

1) Since the L2 creation flow was changed to run on both hosts
simultaneously, one future now handles two devices. This means that
iterating over the futures can yield two hostnames instead of one.
This breaks the octavia_l2_failed_futures metric, which only takes one
hostname as label.
This commit fixes this by treating all values of the futures
dictionary as lists and (in the case of a failure) iterating over them
to set the octavia_l2_failed_futures metric for each host therein.

2) After the futures have been iterated over, the failed BigIPs are
checked to see whether all BigIP hosts/guest flows failed. This check
needs BigIPRestClient instances (to compare to self._bigips and
self.vcmps, respectively), but for the BigIP guests only the hostnames
are provided in the futures list.

velp commented on octavia_f5/tests/unit/controller/worker/test_l2_sync_manager.py:
```diff
@@ -100,12 +100,12 @@ def test_ensure_l2_flow_all_available(self, mock_get_network,
         self.manager.ensure_l2_flow(mocked_selfips, 'test-network-id')
         mock_l2_flow.assert_called_once_with(data={
-            'test-guest-hostname_0': {
+            self.manager._bigips[0]: {
```
How is this change related to this PR? I would ask you to keep it away (especially if it is not changing anything).

The same for all other lines in this file.

BenjaminLudwigSAP replied:
This change is necessary. The keys of the `data` dictionary don't contain the BigIP hostnames anymore, but the `BigIPRestClient` objects themselves. So these tests fail if the names aren't replaced with references to those objects.

velp replied:
Ok, got it. But anyway, I pushed a stacked PR that does not require these changes.

BenjaminLudwigSAP replied:
Fixed in new commit, please review.

velp commented on octavia_f5/controller/worker/l2_sync_manager.py:
```diff
@@ -202,13 +202,13 @@ def ensure_l2_flow(self, selfips: [network_models.Port], network_id: str, device
         fs[self.executor.submit(
             self._do_ensure_l2_flow,
-            data=ensure_l2_flow_data)] = ','.join(ensure_l2_flow_data.keys())
+            data=ensure_l2_flow_data)] = ensure_l2_flow_data.keys()
```
I would rather use the additional list to save filtered bigips objects than putting objects into the dict keys, which is usually a bad idea because `__hash__` and `__eq__` functions are not implemented in these objects correctly, but in this specific case when we are not using keys anywhere else - it's OK.

BenjaminLudwigSAP replied:
Yeah, I also find it a bit icky, but I consciously decided against refactoring that issue in this PR to get this fix through first. It's definitely a good issue for refactoring in the future.
If you can come up with a quick way to refactor it with a very small diff, feel free to make another stacked PR.

velp replied:
Done: https://github.com/sapcc/octavia-f5-provider-driver/pull/292

BenjaminLudwigSAP replied:
Fixed in new commit instead, please review.

velp replied:
This solution is working for me as well. Approved

BenjaminLudwigSAP committed:
[l2] Replace ensure_l2_flow data dictionary with list

A list suffices because the keys of the dictionary aren't needed:

- Since the ensure_l2_flow data contains data for every bigip in
  `self._bigips`, we can simply add `self._bigips` to the futures
  dictionary, instead of `ensure_l2_flow_data.keys()`.

- `_do_ensure_l2_flow` doesn't need a dictionary either, since it only
  uses the values, ignoring the keys. (Just as before, it takes the
  `BigIPRestClient` instance out of the store, which is in the
  dictionaries that are the items of the list)

Closes #292
